core/src/main/scala/kafka/zk/AdminZkClient.scala

```diff
@@ -365,6 +365,29 @@ class AdminZkClient(zkClient: KafkaZkClient) extends Logging {
       case ConfigType.Ip => changeIpConfig(entityName, configs)
       case _ => throw new IllegalArgumentException(s"$entityType is not a known entityType. Should be one of ${ConfigType.all}")
     }
+
+    if ((ConfigType.Client.equals(entityType) || ConfigType.User.equals(entityType) || ConfigType.Ip.equals(entityType)) && configs.isEmpty) {
+      val currPath = ConfigEntityZNode.path(entityType, entityName)
+      if (zkClient.getChildren(currPath).isEmpty) {
+        var pathToDelete = currPath
+        if (isUserClientId) {
+          val user = entityName.substring(0, entityName.indexOf("/"))
+          val clientId = entityName.substring(entityName.lastIndexOf("/") + 1)
+          val clientsPath = ConfigEntityZNode.path(ConfigType.User, user + "/" + ConfigType.Client)
+          val clientsChildren = zkClient.getChildren(clientsPath)
+          if (clientsChildren.size == 1 && clientsChildren.head.equals(clientId)) {
+            pathToDelete = clientsPath
+            val userData = fetchEntityConfig(ConfigType.User, user)
+            val userPath = ConfigEntityZNode.path(ConfigType.User, user)
+            val userChildren = zkClient.getChildren(userPath)
+            if (userData.isEmpty && userChildren.size == 1 && userChildren.head.equals(ConfigType.Client)) {
+              pathToDelete = userPath
+            }
+          }
+        }
+        zkClient.deletePath(pathToDelete)
```

Comment:
Looking at this block of code, there are apparently three different scenarios for the value of `pathToDelete`, from lines:

1. 372 `var pathToDelete = currPath`
2. 379 `pathToDelete = clientsPath`
3. 384 `pathToDelete = userPath`

 But following through the test cases we only seem to hit two of those. `testChangeConfigsWithUserAndClientId` hits 3. and all the other current tests stick to 1. Should we have an extra test case to cover scenario 2. or, if that scenario should never happen, should this logic be re-structured?

Reply:
Right. Scenario 2 will happen if the user configs are not empty but clients under the user is empty. I have added test case to cover this scenario. Thank you.